### PR TITLE
Intelligent angle searcher + StrategyFromPaper2

### PR DESCRIPTION
**Pull request by Trampeltier**
Added 2 Strategies

### CODE REVIEW

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+import org.locationtech.jts.geom.LineString;
+import org.locationtech.jts.geom.Point;
+import org.locationtech.jts.geom.impl.CoordinateArraySequence;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * I tried here to implement the TreasureHunt2 Algorithm,
+ * which for angles beta < 2 * PI has traveling time in O(n ^ (2 - epsilon)).
```
Comment (jotoh98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java]:
```suggestion
 * which for angles beta &lt; 2 * PI has traveling time in O(n ^ (2 - epsilon)).
```
We have to format the javadoc comments in proper HTML for the CI javadoc compilation to work.

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/StrategyFromPaper.java [hunk 2]
```diff
@@ -232,38 +232,39 @@ private void setRectToPhase() {
         D = JTSUtils.createPoint(startX - halfDiff, startY - halfDiff);
     }
 
-    private Movement rectangleScan(Point A, Point B, Point C, Point D) {
+    public static Movement rectangleScan(Point A, Point B, Point C, Point D) {
         Movement movements = new Movement();
 
         int k = (int) A.distance(B);
```
Comment (jotoh98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/StrategyFromPaper.java]:
```suggestion
        int roundedDistance = (int) A.distance(B);
```
Idk something more descriptive. Refactoring the code (or even just renaming vars) would make it more readable and accelerate the merge and review times a lot.

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/StrategyFromPaper.java [hunk 3]
```diff
@@ -232,38 +232,39 @@ private void setRectToPhase() {
...
+            double xDist = C.getX() - D.getX();
+            double yDist = C.getY() - D.getY();
             for (int i = 0; i <= k; i++) {
-                b[i] = JTSUtils.createPoint(D.getX() + xDist * ((double) i / k), C.getX() + yDist * ((double) i / k));
+                b[i] = JTSUtils.createPoint(D.getX() + xDist * ((double) i / k), D.getY() + yDist * ((double) i / k));
+                //System.out.println("b[ "+ i + "] = " + b[i]);
             }
         }
-
         if (k % 2 == 1) //code like in paper
```
Comment (jotoh98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/StrategyFromPaper.java]:
```suggestion
        if (k % 2 == 1)
```
No inline-comments, extract a method and rename it comprehensible.

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/Tile.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+class Tile {
+    enum Color {black, white};
+
+    /* Corner Positions:
+     * A   B
+     * D   C
+     * */
+    @Getter
+    private Coordinate location;
```
Comment (jotoh98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/Tile.java]: How are the four corner positions represented here? These doc-comments have to be improved.

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/Tile.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    @Getter
+    private int size;
+    @Getter
+    private Color color;
+
+    /*
+        0 1
+        2 3
+     */
+    Tile[] subtiles;
```
Comment (jotoh98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/Tile.java]: Please improve the doc-comments.

**Review by jotoh98 — CHANGES_REQUESTED** [review]
Assuming your code is running, you just have to improve the doc-comments and the readability of your code. Variables named `i` or `k` and huge functions slow down the understanding and due to that the review of your PRs. You also didn't provide any UnitTests, which is bad practice but can be ignored for now. The build of your PR failed because of malformed HTML - I pointed out the exact location.

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/StrategyFromPaper.java [hunk 6]
```diff
@@ -232,47 +232,50 @@ private void setRectToPhase() {
...
-        { //create a_i on line segment AB
+        { 
+            //Create a_i on line segment AB
             double xDist = B.getX() - A.getX();
             double yDist = B.getY() - A.getY();
-            for (int i = 0; i <= k; i++) {
-                a[i] = JTSUtils.createPoint(A.getX() + xDist * ((double) i / k), B.getX() + yDist * ((double) i / k));
+            for (int i = 0; i <= roundedDistance; i++) {
+                a[i] = JTSUtils.createPoint(A.getX() + xDist * ((double) i / roundedDistance), A.getY() + yDist * ((double) i / roundedDistance));
+                //System.out.println("a[ "+ i + "] = " + a[i]);
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/StrategyFromPaper.java]: remove comments we have git

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+        }
+        public void run(){
+            strategy.TreasureHunt2();
+        }
+    }
+
+    /*Basically, function from paper.*/
+    private void TreasureHunt2(){
+        /*for (int i = 1; i <= 360; i++){
+            System.out.println("Index( " + i + ") = "+  index(Angle.toRadians(i)));
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java]: No commented out code

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+
+    /*Basically, function from paper.*/
+    private int Mosaic(int i , int k) {
+        Coordinate O = location.copy();
+        int length = (int)Math.pow(2, i);
+        Tile S = new Tile(new Coordinate(O.getX()-length/2., O.getY()+length/2.), length, nextMoves);
+
+        IndexMax = k;
+        int maxSteps = calcMaxSteps(i, k);
+        //System.out.println("maxSteps: " + maxSteps);
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java]: ...

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+                //Wait for Hint in main thread
+                synchronized (syncObj){
+                    syncObj.notify();
+                    flagMoveIsSet = true;
+                }
+                synchronized (thread) {
+                    try {
+                        thread.wait();
+                    } catch (InterruptedException e) {
+                        //System.out.println("Error waiting for hint in Mosaic().");
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java]: Handle excption properly

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+
+
+    @Override
+    public Movement move() {
+        synchronized (syncObj) {
+            if (!flagMoveIsSet) {
+                try {
+                    syncObj.wait();
+                } catch (InterruptedException e) {
+                    System.out.println("Error waiting in move().");
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java]: No system out print lines we have a logger

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+
+        //Wait for Movement in strategy thread
+        synchronized (thread) {
+            thread.notify();
+        }
+        synchronized (syncObj) {
+            try {
+                syncObj.wait();
+            } catch (InterruptedException e) {
+                System.out.println("Error waiting in move().");
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java]: ...

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+        nextMoves.addAdditionalItem(new GeometryItem<>(new LineString(new CoordinateArraySequence(coordinates),
+                JTSUtils.GEOMETRY_FACTORY), GeometryType.SEARCHER_MOVEMENT));
+        location = c.copy();
+    }
+
+
+    private static int calcMaxSteps(int i, int k){
+        /*Equivalent to formula from paper:
+        Math.ceil(Math.log(Math.sqrt(Math.pow(2, i))) / (Math.log(Math.pow(4,k))))*/
+        //return (int)Math.ceil(0.25*i/(double)k);
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java]: ...

**Review by 1-alex98 — CHANGES_REQUESTED** [review]

**Review by 1-alex98 — DISMISSED** [review]

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+                //Wait for Hint in main thread
+                synchronized (syncObj){
+                    syncObj.notify();
+                    flagMoveIsSet = true;
+                }
+                synchronized (thread) {
+                    try {
+                        thread.wait();
+                    } catch (InterruptedException e) {
+                        //System.out.println("Error waiting for hint in Mosaic().");
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java]: ...

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+        /*The formula from the paper is:
+        Math.ceil(Math.log(Math.sqrt(Math.pow(2, i))) / (Math.log(Math.pow(4,k))))
+        which is equivalent to (int)Math.ceil(0.25*i/(double)k),
+        but the custom modified formula works a tiny bit better:*/
+        return (int)Math.ceil(i/(double)k);
+    }
+
+    
+    private static int index(double alpha){
+        /*The index function
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java]: java doc

File: src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+        nextMoves.addWayPoint(JTSUtils.GEOMETRY_FACTORY.createPoint(c));
+        Coordinate[] coordinates = {location, c};
+        nextMoves.addAdditionalItem(new GeometryItem<>(new LineString(new CoordinateArraySequence(coordinates),
+                JTSUtils.GEOMETRY_FACTORY), GeometryType.SEARCHER_MOVEMENT));
+        location = c.copy();
+    }
+
+
+    private static int calcMaxSteps(int i, int k){
+        /*The formula from the paper is:
```
Comment (1-alex98) [inline src/main/java/com/treasure/hunt/strategy/searcher/impl/sfp2/StrategyFromPaper2.java]: java doc

**Review by 1-alex98 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**